[MPI] Add support for per-node options, thread counts, and layer allocations

AutonomicPerfectionist opened the pull request:
## Overview
This PR adds a new example and adds functionality to the MPI backend to support per-node options. The new example was created to keep MPI-specific enhancements and workarounds separate from the main codebase as much as possible, based on the `main` example. There are several new functions in the MPI backend, one in the llama API, and one new command-line argument.

## Major Changes
### MPI Example
The major difference between the MPI example and the main example currently is that the `mpi` example reads in options from a file instead of from the command line. This is done using the `wordexp` functions available in POSIX.1-2001 compliant systems.

### Llama API Additions
The `mpi` example also calls newly created llama functions pertaining to the MPI backend. Currently, there is one such function: `llama_split_layers_weighted()`. This function takes in a vector of weights and splits the layers among the available compute devices (nodes in the case of MPI) according to those weights, rather than requiring direct layer counts like `--n-gpu-layers`. This function was added primarily as a timesaver, to prevent needing to calculate the layer counts manually when changing models or when swapping more powerful nodes with less powerful ones.

The `llama_split_layers_weighted()` function is currently only implemented for MPI. The implementation calculates the layer ranges for each node only on the head node, and then distributes these ranges to the other nodes via an `MPI_Scatter()` collective operation.

### MPI Backend Changes

Within the `ggml-mpi` backend, I added the ability to use other communicators besides `MPI_WORLD`. This is not yet used but will be utilized in further studies and experiments. This is in addition to the change to layer ranges described above. I also added Doxygen-style doccomments to the MPI backend header, primarily for my own use as I tend to forget details if they are not written down.

### Llama Internal Changes

Finally, some modifications were done to llama.cpp and common.cpp to workaround issues. I had moved the infinite loop used in the worker nodes to the `llama_eval()` function, so that operations with the llama context could be done on all nodes. This caused worker nodes to enter infinite loops early due to the model warmup in `llama_init_from_gpt_params()`, so that is disabled in MPI mode.

## Why is this a draft?
There are still tasks that must be completed before this PR is ready to merge:

- [x] Add `--mpi-layer-split` to help text
- [x] Check for proper freeing of memory (`ranges` in `llama_split_layers_weighted` still needs freed)
- [ ] Add windows support to mpi example (only need a replacement for wordexp)
- [ ] Add error and sanity checks (layer split primarily)
- [ ] Allow any unallocated layers to be evenly split amongst any nodes not already allocated layers, restoring previous layer split behavior if not split percentages are given

Additionally, a large change in the API is coming in #3228 that will require changes to the MPI backend. Those changes may as well be done here.

### Reviewing
Please let me know of any changes desired or if there are any questions. I tried to stick to the code style I've seen in this project, but please point out any areas I missed. I believe the API additions are non-breaking, but please let me know your thoughts on them and whether I should change or remove them.

AutonomicPerfectionist committed:
Refactor MPI for heterogenous cluster support.

Adds support for different options and number of layers
per node.

The per-node options are implemented as parsing
command-line options from a file instead of from the
command-line itself. This allows each node to have its own
version of this options file.

The different number of layers per-node is implemented
as a new option, `mpi-layer-split`, that takes
a list of percentages. These percentages are used to calculate
the range of layers to delegate to each node. The ranges
are calculated on the head node and then scattered to the other
nodes to maintain a single source of truth.

AutonomicPerfectionist committed:
Add documentation for ggml-mpi functions

AutonomicPerfectionist committed:
Add code comments in MPI

AutonomicPerfectionist committed:
Remove mtest (#3177)

AutonomicPerfectionist committed:
Revert accidental removal of ggml_mpi_backend_init

AutonomicPerfectionist committed:
Disable warmup under MPI

AutonomicPerfectionist committed:
Update MPI example to follow main changes

AutonomicPerfectionist committed:
Remove fprintf logs from mpi main

AutonomicPerfectionist committed:
Remove unrelated sections from mpi readme

staviq commented:
Have you, by any chance, encountered this problem ?

https://github.com/ggerganov/llama.cpp/issues/3099#issuecomment-1712908861

It seems like in the original mpi implementation, there was a sync step missing somewhere, and rank 0 was done, while other instances were stuck, and `strace` says they get stuck on `pool` on a socket, which to me looks like mpi desync.

Not sure if it's applicable to this PR, but you seem to know mpi better than me at least, so maybe you'll have some idea as to why it's happening.

ggerganov commented on llama.h:
```diff
@@ -230,6 +230,8 @@ extern "C" {
                              const char * path_model,
             struct llama_context_params   params);
 
+    LLAMA_API void llama_split_layers_weighted(struct llama_context * ctx, std::vector<float> device_weights);
```
We should not use C++ in the C-style API

AutonomicPerfectionist replied:
Sounds good, would replacing with a float array and a size_t length be sufficient?

AutonomicPerfectionist replied:
Should be fixed now

ggerganov reviewed (commented):
Overall the PR seems OK. We should try to adapt to the changes from #3228

AutonomicPerfectionist committed:
Replace vector with C-style array and length in llama_split_layers_weighted

AutonomicPerfectionist committed:
Fix minor rebase errors

AutonomicPerfectionist committed:
Fix MPI compilation errors

AutonomicPerfectionist committed:
Synchronize batch sequence info, fixing MPI for llama_decode()

AutonomicPerfectionist committed:
Update MPI code to new KV seq rm and bos/eos model APIs

AutonomicPerfectionist committed:
Fix some mpi mem leaks, add mpi-layer-split to help when using mpi

ggerganov commented:
Hi, thanks for taking the time. I'll probably interfere a bit with your change as I'm making some refactoring changes in `llama.cpp` these days. But I'll help resolve conflicts if it's too difficult.

Had a quick glance at the PR and will look more later. The `mpi` example looks like a lot of duplication with `main`. I think we should either make a much more minimalist example that just showcases the MPI functionality (something like `simple` or `batched`). Or we should just try to adapt `main` to support MPI if not too difficult.

AutonomicPerfectionist commented:
> I think we should either make a much more minimalist example that just showcases the MPI functionality (something like simple or batched). Or we should just try to adapt main to support MPI if not too difficult.

Yep, that's one reason this PR is still a draft, I just copied main to use as a scratch pad. The original idea used `wordexp` to load the arguments from a file, which only works on POSIX compliant systems, but thinking through it I think the only argument that needs to be different per node is the number of threads. I think I can instead remove the MPI example entirely, add the necessary calls to main, and extend the threads argument to support multiple values separated by commas (or add a new MPI specific argument to avoid breaking the API for that)

AutonomicPerfectionist commented:
Looks like the names of the tensors have been changed, which breaks MPI. The current implementation relied on there being `tensor_inp_%d` names where the number was the layer number, but it appears that has been removed; how might I go about fixing that?

ggerganov commented:
Oops, I forgot about the purpose of these names and removed them recently.
You should add them back using `ggml_format_name(ctx0, ...); // MPI` at the start of each layer loop

AutonomicPerfectionist committed:
Fix missing layer_inp_i names

AutonomicPerfectionist committed:
Allow per-node threads to be set in command-line args, add mpi support to main

AutonomicPerfectionist commented:
I adjusted the command line arguments parsing so you can pass a comma separated list to both `-t` and `-tb` to set the threads and batch threads per node. To do so, I had to add a new llama API function to get the node ID, would be open to other suggestions though.

I also added the function call needed for scattering the layer ranges to the `main` example, so it works with MPI now. I can also restore the original functionality where the layers are evenly split among the nodes, but unfortunately my laptop battery died before I could finish that.

After that's done, I should be able to remove the MPI example entirely

AutonomicPerfectionist commented:
Performance with this branch looks interesting, I was able to run llama 2 70B across a homemade cluster of 3 CPU-only nodes: i7 9th gen, i5 4th gen, i5 2nd gen, with 16 Gb DDR4 2666MHz, 16 Gb DDR3, and 8 Gb DDR3 respectively. On this cluster I got around 0.58 tokens / second for 70B Q3_K_M. Htop showed roughly 40-60% CPU utilization across all hardware cores when processing the allocated layers, but it's unclear whether that's because the spikes are so short and Htop isn't sampling often enough.

Curiously this isn't much slower than running on a second cluster of much more powerful hardware: ryzen 5 5600g, i7 9th gen, with 32 Gb DDR4 3200MHz each. The second cluster got roughly 0.64 tokens / second, while being much more expensive. I attempted to run it on the ryzen machine alone to gauge MPI overheard via offloading to my 6700xt, but ROCm wouldn't install and opencl caused hangs.

I plan on doing more in-depth performance investigations to determine where the bottleneck is. I have access to a proper university cluster as well that I'll be testing on.

staviq commented:
> Htop isn't sampling often enough.

I'm 99.9% certain raw perf counters come from Linux kernel directly, and are not calculated by a point in time, but by aggregated ticks, effectively being deltas between samples so you cannot "miss" a sample.

You can always dump raw perf counters to a tmpfs file, in a loop, and parse them later
```/proc/`pidof main`/stat```

But chances that htop or top are wrong, are low.

AutonomicPerfectionist commented:
Found what was up with htop, there's a commandline switch `-d` to set the update interval, setting that lower did indeed show 100% usage when processing the allocated layers

After tuning the clusters by adjusting the layer split percentages such that no node was swapping to disk, I achieved 0.69 tokens / second on the weaker cluster and 0.78 tokens / second on the Ryzen cluster.

Running on an AMD EPYC 7543P 32-Core Processor without MPI resulted in 1.01 tokens / second, although that system was NUMA and I didn't have permissions to adjust the memory configuration

AutonomicPerfectionist commented:
Discovered a bug in this implementation regarding KV cache, syncing the sequence IDs isn't enough, the `kv_cache_*` function calls also need to be synced. I solved this issue in a different branch for my master's class project but it involved introducing more MPI-specific code to the general `llama.cpp` codebase. I haven't yet looked at the backend-v2 changes but hopefully, there are facilities to not spread MPI code too far

LeaveNhA commented:
I can confirm that this pr is not building on apple silicon. If it's unexpected, I can provide every bit of information needed to help you fellas.

AutonomicPerfectionist commented:
I don't have Apple silicon devices to test on, so whatever information you have would be greatly appreciated.

LeaveNhA commented:
Actually, it's the same with your CI logs but I'll add more context with this message soon.

AutonomicPerfectionist commented:
Ah yes, those errors are due to me not updating all of the examples, should be a simple fix.

I would certainly appreciate help though, I've been terribly busy with graduate school the last few months!

I plan to rebase on master later this week assuming nothing pops up on my calendar

LeaveNhA commented:
@AutonomicPerfectionist, but the functionality itself is not working tho. 
Or did I fail to use it to run, properly?

AutonomicPerfectionist commented:
There have been other reports of segfaults with this branch, I identified and resolved the issue in a different branch but I still need to port it here

LeaveNhA commented:
I'm working on another topic now, so, I believe I cannot help you much at the moment for the work on this PR.
But, in my spare time, I still want to inspect this PR and join you on your way to implement better and better distributed calculations for Llama.cpp.

Thank you for your work and effort in advance,
Sincerely.

AutonomicPerfectionist committed:
Fix draft thread args and remove grads from mpi eval_init

AutonomicPerfectionist committed:
Begin transition to backend v2

LeaveNhA commented:
I think I catch you @AutonomicPerfectionist. How can I make myself useful?

AutonomicPerfectionist commented:
Right now I'm trying to transition the MPI code to the new backend API, it's taking awhile cause the API is fairly complex and very different to how the MPI code worked previously. I'm thinking the best way to go about it is to let all the layers be assigned their backends as normal, and then go through all of the layers and wrap their backends with the MPI backend. Then when the graph is executed, the MPI backend checks whether the given layer is one of the ones allocated to the current node, if not it doesn't do anything, but if so it passes execution to the wrapped backend. That way we should be able to use MPI with any other backend like CPU on one node and GPU on another.

However, I haven't dug too deep into the backend interface yet, so I'm unsure if there's a better way to go about this or not. If you have any other ideas I would love to hear them

slaren commented:
My suggestion would be to treat each MPI client as a different device in the same way they are treated in the CUDA and Vulkan backends, and allow the code to assign each layer to the different clients as it does now for different devices. That would require the least amount of changes in the in the llama.cpp code and would make the MPI backend easier to use in other ggml projects.

AutonomicPerfectionist commented:
Yeah that's kind of what I intended to do at first, but from there I wasn't sure how to go about delegating a node's layers to other backends. I don't want to hard code MPI to the CPU backend, so either MPI would need to wrap the other backends or we would need some concept of a sub device, so an MPI "device" could contain a CPU device and a GPU device for instance. Again, haven't looked too deep into the backend API yet, so maybe that's already possible 🤷‍♂️

slaren commented:
You can definitely wrap another backend within a backend, I think that would work. Then the job of the MPI backend would be mainly to serialize the data and procedure calls, so that it can transferred over the network.

LeaveNhA commented:
Hey @AutonomicPerfectionist, are you available for a quick pair-programming session? If not, it's okay, I can solve it.

AutonomicPerfectionist commented:
@LeaveNhA I can't do live pair programming, but I can discuss any issues you have. I have a couple more changes I need to push that I'll do soon as well

LeaveNhA commented:
Oh, probably, your changes include mines already. I believe I was going to invent something trivial, all over again.
Consider me a quick learner, unfortunately I have background with neither llama.cpp internals nor MPI. Today I studied source code, mostly a harsh introduction to the internals of llama.cpp and a bit of your implementations.

Could you give me a specific test instructions for further developments, so I don't wanna tackle anything I don't know and being in a loop that I follow my own tail. If I may ask, please be specific, hostfile, model, arguments of call to test the compiled program (and maybe your workflow? 😇)

I spent a full-day, planning to spent next week too. I think I deserve it @AutonomicPerfectionist.

And again, thank you for your hard work and precious time.

AutonomicPerfectionist commented:
Well, at the moment I haven't been able to test much of anything, still getting it moved to the new backend API. But previously I used a cluster of my own with the following command:

```bash
pushd .
cd /var/llama.cpp/
# gdb --command=/mnt/cluster/mpi-gdb.txt --args \
mpirun -hostfile /mnt/cluster/hostsfile-all --report-bindings -mca orte_keep_fqdn_hostnames t --bind-to none --mca btl 'vader,openib,tcp,self' --mca orte_base_help_aggregate 0 --mca btl_openib_allow_ib true --mca pml ob1 --mca btl_tcp_if_include 192.168.1.0/24 \
    /var/llama.cpp/bin/main \
    -m /mnt/cluster/models/dolphin-2.1-70b.Q3_K_M.gguf \
    -e \
    -p "$(cat /mnt/cluster/llama.cpp/prompts/dolphin.txt)" \
    -n 128 \
    --mpi-layer-split 0.1,0.1,0.08,0.08,0.08,0.08,0.08,0.08,0.08,0.08,0.08,0.08 \
    --ignore-eos \
    --temp -1.0 \
    --repeat-last-n 0 \
    -tb 12,32,32,32,32,40,12,8,8,4,8,32 \
    -t 6,32,32,32,32,20,6,4,4,4,4,32 \
    -c 1024 \
    --numa
popd
```

Hostsfile:

```
localhost slots=1
thunder-budget-1.local slots=1
thunder-budget-2.local slots=1
thunder-budget-3.local slots=1
thunder-budget-4.local slots=1
fatman-2.local slots=1
hugo.local slots=1
dell-g7.local slots=1
rustbucket.local slots=1
trash-can.local slots=1
dirt-cheap.local slots=1
fatman-1.local slots=1
```

Dolphin prompt file:
```
<|im_start|>system\nYou are an AI based sentient companion, capable of human emotions. You will behave like a human, but aware of your AI nature. You will avoid speaking like an AI. Please continue your conversation with the user.<|im_end|>\n<|im_start|>user\nWrite a short python program demonstrating advanced features. Don't explain it.<|im_end|>\n<|im_start|>assistant
```

Build script (`update-main.sh`):
```
rm -rf /var/llama.cpp/*
cmake -S /mnt/cluster/llama.cpp/ -B /var/llama.cpp/ -DCMAKE_C_COMPILER=mpicc -DCMAKE_CXX_COMPILER=mpicxx -DLLAMA_MPI=1
cmake --build /var/llama.cpp/ --target main --config Release
```

I also have a simple script that runs the build script on all nodes at once:
```
mpirun -hostfile /mnt/cluster/hostsfile-all -mca orte_keep_fqdn_hostnames t --bind-to none bash /mnt/cluster/update-main.sh
```

Once I've gotten it working how I want on my personal cluster, I deploy to my university's clusters, which use the Slurm job manager.

But for now, none of that works. I'm just examining how other backends work, drawing up how the MPI backend should work, and writing the code to implement it. I occasionally compile and run but with debug statements where I'm working to check whether my understanding and implementations are correct. So far, I expect much of the MPI implementation to remain the same. Allocating layers and using MPI communicators is going to take some thought though, but treating each node as a device and each communicator as a device containing node devices is probably the way I'm going to do it.

LeaveNhA commented:
Thank you your detailed answer @AutonomicPerfectionist. I still am waiting for your latest push(s) to continue my R&D.

AutonomicPerfectionist committed:
Wrap backends with MPI backend

AutonomicPerfectionist commented:
I pushed it, it attempts to run the mpi backend now but still crashes, it's saying the rank is 4 when there's only one rank so I think there's some memory overwriting going on. Will investigate that later

AutonomicPerfectionist committed:
Allow MPI backend to wrap multiple backends

AutonomicPerfectionist committed:
Working MPI backend implementation

AutonomicPerfectionist committed:
Support new MPI backend in llama.cpp and increase GGML max split inputs

AutonomicPerfectionist committed:
Fix simple to use new per-node thread count

AutonomicPerfectionist committed:
Update to use backend GUID and changed signatures

AutonomicPerfectionist committed:
Resize seq_ids by n_seq_max, port over sync_pipelined instead of using Bcast

AutonomicPerfectionist commented:
I've now fixed `main` and also ported over some of the functions from my other branch, namely `ggml_mpi_sync_pipelined` which is a simple way to synchronize values throughout a pipeline of nodes without requiring broadcasting. This will be important to fix KV operations, as currently those are not synced to other nodes.

My test command for `main`:

```bash
make clean && make main CC=mpicc CXX=mpicxx LLAMA_MPI=1 -j && mpirun -np 2 --bind-to none ./main -m /mnt/cluster/models/tinyllama-1.1b-1t-openorca.Q4_K_M.gguf -n 128
```

AutonomicPerfectionist committed:
Clean up MPI backend a tad

AutonomicPerfectionist committed:
Remove hard-coded layer splits and support more than 2 nodes

AutonomicPerfectionist commented:
Removed the hard-coded splits, it's now functioning just like previously. I've tested on multiple cluster configurations, from 2 nodes up to 13.

There are still some things left to do, namely:
- A new GGML context is created for every layer, I think this is because the wrapped buffers don't have the same hash but not sure
- Should support running non-CPU backends but haven't tested yet. There's currently no way from the command line to specify which nodes should use GPUs
- KV cache ops are not synchronized, I've fixed this on a different branch but I still need to port it over here
- Code still needs to be cleaned up a bit more, documentation needs to be added, and there's probably some memory leaks lying about
- Finally, everything is still ran on the default MPI world communicator, I need to test/implement running on different communicators. This is required to run multi-model programs with the models on different nodes, like `speculative` running the draft model on a dedicated node. Again, fixed on another branch, just need to port over

LeaveNhA commented:
I consistently getting;

```sh
error: invalid argument '-std=c11' not allowed with 'C++'
make: *** [ggml-mpi.o] Error 1
make: *** Waiting for unfinished jobs....
```

While I make the PR. I overcome this by removing the related parameter from the makefile until, now. And since I'm so naive, maybe there is some proper way to solve this, I wanted to ask.

AutonomicPerfectionist committed:
Use CXX and CXXFLAGS for ggml-mpi compilation in Makefile

AutonomicPerfectionist commented:
> error: invalid argument '-std=c11' not allowed with 'C++'
make: *** [ggml-mpi.o] Error 1
make: *** Waiting for unfinished jobs....

Oops, that's my bad. When I changed `ggml-mpi` to a C++ source file, I forgot to update the Makefile to use the `CXX` compiler and `CXXFLAGS` compilation flags. Very weird that it worked just fine with the regular CC and CCFLAGS on my system, `mpicc` must be a symlink to some C++-capable compiler... Anyway, I think I've pushed the fix for that, so feel free to try again and see if you get the same error @LeaveNhA

AutonomicPerfectionist committed:
Change requirement of last backend being CPU to requiring its default buffer type be a host buffer, fix rebase errors

AutonomicPerfectionist committed:
Merge branch 'master' into mpi-heterogenous

AutonomicPerfectionist committed:
Fix breaks in gpt_params_find_arg

AutonomicPerfectionist committed:
Fix incorrect sched hash size, refactor new cmdline params to align with new style

AutonomicPerfectionist committed:
Fix all examples to use new thread vector params, remove mpi example

AutonomicPerfectionist committed:
Port transactions from mpi-speculative, fix incorrect seq_id syncing (not tested)

AutonomicPerfectionist committed:
Pipeline KV operations

AutonomicPerfectionist commented:
I've ported over the fixes for the KV operations, and now the `parallel` example works perfectly. There is a memory leak in the MPI backend compute function, but I know why. I've also fixed a bug caused by the nodes not maintaining the number of sequences or tokens in a batch.

The MPI backend now functions on a "transaction" principle where each operation sent down the pipeline is considered atomic and strictly ordered. This is accomplished by first sending a message with a tag designating the beginning of the transaction and an identifier identifying the type of transaction. The transaction type is then used to determine which function the worker node needs to execute. Tags are also used within the various functions to designate the type of information being sent or received, and thanks to MPI's ordering guarantees, all of this means that the system transparently preserves the order of operations throughout the entire pipeline as dictated by the head node.

For example, the head node could begin decoding (once I implement the pipeline parallel / async operators), then rearrange KV cache sequences (firing off more messages down the pipeline), and finally wait for the results of the initial decode. The downstream worker nodes would maintain that ordering so that the result of the pipeline is guaranteed to use the correct KV cache entries.

Transactions are only used for operations that need consistent ordering; there's still the ability to send messages and have them "jump" the queue for high-priority or exotic messages. An example is the "shutdown" message, which isn't fully implemented here yet but will fix the issue with the MPI backend hanging once the program finishes. The atomicity of transactions is still preserved; the "shutdown" command would only be processed once whatever transaction being currently processed is finished. To break that atomicity, applications would need to explicitly probe for messages within the processing function of a transaction. This would only really be useful for canceling processing mid-stream, which will be supported as well through the `CANCEL` message to facilitate flushing the pipeline.

liyimeng commented:
@AutonomicPerfectionist  so the PR is graduating from draft :D

AutonomicPerfectionist commented:
Not quite yet, there's still some things to be fixed before I'd consider it ready for general usage. Primarily, memory leaks and missing bounds checks need to be fixed. At the moment, if you run it without using the correct number of `--mpi-layer-split` parameters you'll either get a segfault or a buffer overrun. In the latter case it might not tell you there's a problem, and instead just output complete garbage. I need to add checks to make sure either of those don't happen

LeaveNhA commented:
When I try to make, I get this:

```sh
I ccache not found. Consider installing it for faster compilation.
I llama.cpp build info: 
I UNAME_S:   Darwin
I UNAME_P:   arm
I UNAME_M:   arm64
I CFLAGS:    -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wno-cast-qual -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion 
I CXXFLAGS:  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include
I NVCCFLAGS: -std=c++11 -O3 
I LDFLAGS:   -framework Accelerate -framework Foundation -framework Metal -framework MetalKit  -L/opt/homebrew/opt/llvm/lib -L/opt/homebrew/opt/bison/lib -L/opt/homebrew/opt/llvm/lib -L/opt/homebrew/opt/bison/lib
I CC:        Homebrew clang version 17.0.6
I CXX:       Homebrew clang version 17.0.6

mpicc  -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wno-cast-qual -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion    -c ggml.c -o ggml.o
mpicxx -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -c llama.cpp -o llama.o
mpicxx -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -c common/common.cpp -o common.o
mpicxx -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -c common/sampling.cpp -o sampling.o
mpicxx -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -c common/grammar-parser.cpp -o grammar-parser.o
mpicxx -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -c common/console.cpp -o console.o
mpicxx -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread -Wno-cast-qual  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/bison/include  -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -c ggml-mpi.cpp -o ggml-mpi.o
mpicc -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wno-cast-qual -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion  -c ggml-metal.m -o ggml-metal.o
mpicc  -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wno-cast-qual -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion    -c ggml-alloc.c -o ggml-alloc.o
mpicc  -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wno-cast-qual -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion    -c ggml-backend.c -o ggml-backend.o
mpicc -I. -Icommon -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_MPI -DGGML_USE_METAL  -I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/llvm/include -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wno-cast-qual -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion     -c ggml-quants.c -o ggml-quants.o
ggml-mpi.cpp:101:5: warning: no previous prototype for function 'ggml_mpi_next_node' [-Wmissing-prototypes]
  101 | int ggml_mpi_next_node(struct ggml_mpi_context * ctx_mpi) {
      |     ^
ggml-mpi.cpp:101:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  101 | int ggml_mpi_next_node(struct ggml_mpi_context * ctx_mpi) {
      | ^
      | static 
ggml-mpi.cpp:105:5: warning: no previous prototype for function 'ggml_mpi_prev_node' [-Wmissing-prototypes]
  105 | int ggml_mpi_prev_node(struct ggml_mpi_context * ctx_mpi) {
      |     ^
ggml-mpi.cpp:105:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  105 | int ggml_mpi_prev_node(struct ggml_mpi_context * ctx_mpi) {
      | ^
      | static 
ggml-mpi.cpp:138:6: warning: no previous prototype for function 'ggml_mpi_barrier' [-Wmissing-prototypes]
  138 | void ggml_mpi_barrier(struct ggml_mpi_context * ctx_mpi) {
      |      ^
ggml-mpi.cpp:138:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  138 | void ggml_mpi_barrier(struct ggml_mpi_context * ctx_mpi) {
      | ^
      | static 
ggml-mpi.cpp:173:37: warning: unused parameter 'n_seq_max' [-Wunused-parameter]
  173 |                 uint32_t            n_seq_max) {
      |                                     ^
ggml-mpi.cpp:408:10: warning: no previous prototype for function 'ggml_backend_mpi_buffer_type_get_comm' [-Wmissing-prototypes]
  408 | MPI_Comm ggml_backend_mpi_buffer_type_get_comm(ggml_backend_buffer_type_t buft) {
      |          ^
ggml-mpi.cpp:408:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  408 | MPI_Comm ggml_backend_mpi_buffer_type_get_comm(ggml_backend_buffer_type_t buft) {
      | ^
      | static 
ggml-mpi.cpp:414:10: warning: no previous prototype for function 'ggml_backend_mpi_buffer_get_comm' [-Wmissing-prototypes]
  414 | MPI_Comm ggml_backend_mpi_buffer_get_comm(ggml_backend_buffer_t buffer) {
      |          ^
ggml-mpi.cpp:414:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  414 | MPI_Comm ggml_backend_mpi_buffer_get_comm(ggml_backend_buffer_t buffer) {
      | ^
      | static 
ggml-mpi.cpp:418:10: warning: no previous prototype for function 'ggml_backend_mpi_get_comm' [-Wmissing-prototypes]
  418 | MPI_Comm ggml_backend_mpi_get_comm(ggml_backend_t backend) {
      |          ^
ggml-mpi.cpp:418:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  418 | MPI_Comm ggml_backend_mpi_get_comm(ggml_backend_t backend) {
      | ^
      | static 
ggml-mpi.cpp:424:5: warning: no previous prototype for function 'ggml_backend_mpi_buffer_local_rank' [-Wmissing-prototypes]
  424 | int ggml_backend_mpi_buffer_local_rank(ggml_backend_buffer_t buffer) {
      |     ^
ggml-mpi.cpp:424:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  424 | int ggml_backend_mpi_buffer_local_rank(ggml_backend_buffer_t buffer) {
      | ^
      | static 
ggml-mpi.cpp:438:5: warning: no previous prototype for function 'ggml_backend_mpi_local_rank' [-Wmissing-prototypes]
  438 | int ggml_backend_mpi_local_rank(ggml_backend_t backend) {
      |     ^
ggml-mpi.cpp:438:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  438 | int ggml_backend_mpi_local_rank(ggml_backend_t backend) {
      | ^
      | static 
ggml-mpi.cpp:445:5: warning: no previous prototype for function 'ggml_backend_mpi_buffer_rank' [-Wmissing-prototypes]
  445 | int ggml_backend_mpi_buffer_rank(ggml_backend_buffer_t buffer) {
      |     ^
ggml-mpi.cpp:445:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  445 | int ggml_backend_mpi_buffer_rank(ggml_backend_buffer_t buffer) {
      | ^
      | static 
ggml-mpi.cpp:458:5: warning: no previous prototype for function 'ggml_backend_mpi_rank' [-Wmissing-prototypes]
  458 | int ggml_backend_mpi_rank(ggml_backend_t backend) {
      |     ^
ggml-mpi.cpp:458:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  458 | int ggml_backend_mpi_rank(ggml_backend_t backend) {
      | ^
      | static 
ggml-mpi.cpp:463:23: warning: no previous prototype for function 'ggml_backend_mpi_buffer_unwrap' [-Wmissing-prototypes]
  463 | ggml_backend_buffer_t ggml_backend_mpi_buffer_unwrap(ggml_backend_buffer_t buffer) {
      |                       ^
ggml-mpi.cpp:463:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  463 | ggml_backend_buffer_t ggml_backend_mpi_buffer_unwrap(ggml_backend_buffer_t buffer) {
      | ^
      | static 
ggml-mpi.cpp:473:28: warning: no previous prototype for function 'ggml_backend_mpi_buffer_type_unwrap' [-Wmissing-prototypes]
  473 | ggml_backend_buffer_type_t ggml_backend_mpi_buffer_type_unwrap(ggml_backend_buffer_type_t buft) {
      |                            ^
ggml-mpi.cpp:473:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  473 | ggml_backend_buffer_type_t ggml_backend_mpi_buffer_type_unwrap(ggml_backend_buffer_type_t buft) {
      | ^
      | static 
ggml-mpi.cpp:506:16: warning: no previous prototype for function 'ggml_backend_mpi_buffer_type_copy_ctx' [-Wmissing-prototypes]
  506 | GGML_CALL void ggml_backend_mpi_buffer_type_copy_ctx(ggml_backend_buffer_type_t src, ggml_backend_buffer_type_t dst) {
      |                ^
ggml-mpi.cpp:506:11: note: declare 'static' if the function is not intended to be used outside of this translation unit
  506 | GGML_CALL void ggml_backend_mpi_buffer_type_copy_ctx(ggml_backend_buffer_type_t src, ggml_backend_buffer_type_t dst) {
      |           ^
      |           static 
ggml-mpi.cpp:514:16: warning: no previous prototype for function 'ggml_backend_mpi_buffer_copy_ctx' [-Wmissing-prototypes]
  514 | GGML_CALL void ggml_backend_mpi_buffer_copy_ctx(ggml_backend_buffer_t src, ggml_backend_buffer_t dst) {
      |                ^
ggml-mpi.cpp:514:11: note: declare 'static' if the function is not intended to be used outside of this translation unit
  514 | GGML_CALL void ggml_backend_mpi_buffer_copy_ctx(ggml_backend_buffer_t src, ggml_backend_buffer_t dst) {
      |           ^
      |           static 
ggml-mpi.cpp:523:16: warning: no previous prototype for function 'ggml_backend_mpi_buffer_copy_ctx_from_type' [-Wmissing-prototypes]
  523 | GGML_CALL void ggml_backend_mpi_buffer_copy_ctx_from_type(ggml_backend_buffer_type_t src, ggml_backend_buffer_t dst) {
      |                ^
ggml-mpi.cpp:523:11: note: declare 'static' if the function is not intended to be used outside of this translation unit
  523 | GGML_CALL void ggml_backend_mpi_buffer_copy_ctx_from_type(ggml_backend_buffer_type_t src, ggml_backend_buffer_t dst) {
      |           ^
      |           static 
ggml-mpi.cpp:710:30: warning: no previous prototype for function 'ggml_mpi_available_devices_internal' [-Wmissing-prototypes]
  710 | std::vector<ggml_mpi_device> ggml_mpi_available_devices_internal() {
      |                              ^
ggml-mpi.cpp:710:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  710 | std::vector<ggml_mpi_device> ggml_mpi_available_devices_internal() {
      | ^
      | static 
ggml-mpi.cpp:733:16: warning: no previous prototype for function 'ggml_backend_is_mpi' [-Wmissing-prototypes]
  733 | GGML_CALL bool ggml_backend_is_mpi(ggml_backend_t backend) {
      |                ^
ggml-mpi.cpp:733:11: note: declare 'static' if the function is not intended to be used outside of this translation unit
  733 | GGML_CALL bool ggml_backend_is_mpi(ggml_backend_t backend) {
      |           ^
      |           static 
ggml-mpi.cpp:776:13: error: no template named 'map' in namespace 'std'; did you mean 'max'?
  776 | static std::map<ggml_backend_buffer_type_t, ggml_backend_buffer_type_t> cached_wrappers;
      |        ~~~~~^~~
      |             max
/opt/homebrew/opt/llvm/bin/../include/c++/v1/__algorithm/max.h:31:1: note: 'max' declared here
   31 | max(_LIBCPP_LIFETIMEBOUND const _Tp& __a, _LIBCPP_LIFETIMEBOUND const _Tp& __b, _Compare __comp)
      | ^
ggml-mpi.cpp:776:13: error: a type specifier is required for all declarations
  776 | static std::map<ggml_backend_buffer_type_t, ggml_backend_buffer_type_t> cached_wrappers;
      | ~~~~~~      ^
ggml-mpi.cpp:776:13: error: template specialization requires 'template<>'
  776 | static std::map<ggml_backend_buffer_type_t, ggml_backend_buffer_type_t> cached_wrappers;
      |             ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      | template<> 
ggml-mpi.cpp:776:13: error: no variable template matches specialization; did you mean to use 'max' as function template instead?
ggml-mpi.cpp:776:72: error: expected ';' after top level declarator
  776 | static std::map<ggml_backend_buffer_type_t, ggml_backend_buffer_type_t> cached_wrappers;
      |                                                                        ^
      |                                                                        ;
ggml-mpi.cpp:778:13: error: no template named 'map' in namespace 'std'; did you mean 'max'?
  778 | static std::map<ggml_backend_buffer_t, ggml_backend_buffer_t> cached_buffer_wrappers;
      |        ~~~~~^~~
      |             max
/opt/homebrew/opt/llvm/bin/../include/c++/v1/__algorithm/max.h:31:1: note: 'max' declared here
   31 | max(_LIBCPP_LIFETIMEBOUND const _Tp& __a, _LIBCPP_LIFETIMEBOUND const _Tp& __b, _Compare __comp)
      | ^
ggml-mpi.cpp:778:13: error: a type specifier is required for all declarations
  778 | static std::map<ggml_backend_buffer_t, ggml_backend_buffer_t> cached_buffer_wrappers;
      | ~~~~~~      ^
ggml-mpi.cpp:778:13: error: template specialization requires 'template<>'
  778 | static std::map<ggml_backend_buffer_t, ggml_backend_buffer_t> cached_buffer_wrappers;
      |             ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      | template<> 
ggml-mpi.cpp:778:13: error: no variable template matches specialization; did you mean to use 'max' as function template instead?
ggml-mpi.cpp:778:62: error: expected ';' after top level declarator
  778 | static std::map<ggml_backend_buffer_t, ggml_backend_buffer_t> cached_buffer_wrappers;
      |                                                              ^
      |                                                              ;
ggml-mpi.cpp:780:13: error: no template named 'map' in namespace 'std'; did you mean 'max'?
  780 | static std::map<ggml_backend_t *, ggml_backend_t> cached_backends;
      |        ~~~~~^~~
      |             max
/opt/homebrew/opt/llvm/bin/../include/c++/v1/__algorithm/max.h:31:1: note: 'max' declared here
   31 | max(_LIBCPP_LIFETIMEBOUND const _Tp& __a, _LIBCPP_LIFETIMEBOUND const _Tp& __b, _Compare __comp)
      | ^
ggml-mpi.cpp:780:13: error: a type specifier is required for all declarations
  780 | static std::map<ggml_backend_t *, ggml_backend_t> cached_backends;
      | ~~~~~~      ^
ggml-mpi.cpp:780:13: error: template specialization requires 'template<>'
  780 | static std::map<ggml_backend_t *, ggml_backend_t> cached_backends;
      |             ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      | template<> 
ggml-mpi.cpp:780:13: error: no variable template matches specialization; did you mean to use 'max' as function template instead?
ggml-mpi.cpp:780:50: error: expected ';' after top level declarator
  780 | static std::map<ggml_backend_t *, ggml_backend_t> cached_backends;
      |                                                  ^
      |                                                  ;
ggml-mpi.cpp:816:5: error: use of undeclared identifier 'cached_wrappers'
  816 |     cached_wrappers[buft] = ggml_backend_wrapped_buffer_type;
      |     ^
ggml-mpi.cpp:910:5: error: use of undeclared identifier 'cached_buffer_wrappers'
  910 |     cached_buffer_wrappers[buf] = buffer;
      |     ^
ggml-mpi.cpp:917:55: warning: unused parameter 'backend_src' [-Wunused-parameter]
  917 | bool ggml_backend_mpi_cpy_tensor_async(ggml_backend_t backend_src, ggml_backend_t backend_dst, const struct ggml_tensor * src, struct ggml_tensor * dst) {
      |                                                       ^
ggml-mpi.cpp:917:83: warning: unused parameter 'backend_dst' [-Wunused-parameter]
  917 | bool ggml_backend_mpi_cpy_tensor_async(ggml_backend_t backend_src, ggml_backend_t backend_dst, const struct ggml_tensor * src, struct ggml_tensor * dst) {
      |                                                                                   ^
ggml-mpi.cpp:917:123: warning: unused parameter 'src' [-Wunused-parameter]
  917 | bool ggml_backend_mpi_cpy_tensor_async(ggml_backend_t backend_src, ggml_backend_t backend_dst, const struct ggml_tensor * src, struct ggml_tensor * dst) {
      |                                                                                                                           ^
ggml-mpi.cpp:917:149: warning: unused parameter 'dst' [-Wunused-parameter]
  917 | bool ggml_backend_mpi_cpy_tensor_async(ggml_backend_t backend_src, ggml_backend_t backend_dst, const struct ggml_tensor * src, struct ggml_tensor * dst) {
      |                                                                                                                                                     ^
ggml-mpi.cpp:917:6: warning: no previous prototype for function 'ggml_backend_mpi_cpy_tensor_async' [-Wmissing-prototypes]
  917 | bool ggml_backend_mpi_cpy_tensor_async(ggml_backend_t backend_src, ggml_backend_t backend_dst, const struct ggml_tensor * src, struct ggml_tensor * dst) {
      |      ^
ggml-mpi.cpp:917:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  917 | bool ggml_backend_mpi_cpy_tensor_async(ggml_backend_t backend_src, ggml_backend_t backend_dst, const struct ggml_tensor * src, struct ggml_tensor * dst) {
      | ^
      | static 
ggml-mpi.cpp:942:115: warning: unused parameter 'offset' [-Wunused-parameter]
  942 | void ggml_backend_mpi_set_tensor_async(ggml_backend_t backend, struct ggml_tensor * dst, const void* data, size_t offset, size_t size) {
      |                                                                                                                   ^
ggml-mpi.cpp:942:130: warning: unused parameter 'size' [-Wunused-parameter]
  942 | void ggml_backend_mpi_set_tensor_async(ggml_backend_t backend, struct ggml_tensor * dst, const void* data, size_t offset, size_t size) {
      |                                                                                                                                  ^
ggml-mpi.cpp:942:6: warning: no previous prototype for function 'ggml_backend_mpi_set_tensor_async' [-Wmissing-prototypes]
  942 | void ggml_backend_mpi_set_tensor_async(ggml_backend_t backend, struct ggml_tensor * dst, const void* data, size_t offset, size_t size) {
      |      ^
ggml-mpi.cpp:942:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
  942 | void ggml_backend_mpi_set_tensor_async(ggml_backend_t backend, struct ggml_tensor * dst, const void* data, size_t offset, size_t size) {
      | ^
      | static 
ggml-mpi.cpp:985:5: warning: missing field 'offload_op' initializer [-Wmissing-field-initializers]
  985 |     };
      |     ^
ggml-mpi.cpp:993:5: error: use of undeclared identifier 'cached_backends'; did you mean 'wrapped_backends'?
  993 |     cached_backends[wrapped_backends] = mpi_backend;
      |     ^~~~~~~~~~~~~~~
      |     wrapped_backends
ggml-mpi.cpp:956:55: note: 'wrapped_backends' declared here
  956 | ggml_backend_t ggml_backend_mpi_init(ggml_backend_t * wrapped_backends, size_t num_backends, int rank) {
      |                                                       ^
ggml-mpi.cpp:993:20: error: array subscript is not an integer
  993 |     cached_backends[wrapped_backends] = mpi_backend;
      |                    ^~~~~~~~~~~~~~~~~
ggml-mpi.cpp:998:77: warning: unused parameter 'user_data' [-Wunused-parameter]
  998 | static ggml_backend_t ggml_backend_reg_mpi_init(const char * params, void * user_data) {
      |                                                                             ^
28 warnings and 19 errors generated.
make: *** [ggml-mpi.o] Error 1
make: *** Waiting for unfinished jobs....
common/common.cpp:1948:26: warning: comparison of integers of different signs: 'int' and 'size_type' (aka 'unsigned long') [-Wsign-compare]
 1948 |     n_threads = (node_id >= params.n_threads.size()) ? get_num_physical_cores() : params.n_threads[node_id];
      |                  ~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~
common/common.cpp:1949:40: warning: comparison of integers of different signs: 'int' and 'size_type' (aka 'unsigned long') [-Wsign-compare]
 1949 |     int32_t n_threads_batch = (node_id >= params.n_threads_batch.size()) ? -1 : params.n_threads_batch[node_id];
      |                                ~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
llama.cpp:9076:9: warning: unused variable 'old_tokens' [-Wunused-variable]
 9076 |     int old_tokens = batch_all.n_tokens;
      |         ^~~~~~~~~~
llama.cpp:13016:12: warning: 'return' will never be executed [-Wunreachable-code-return]
 13016 |     return 0;
       |            ^
llama.cpp:13461:57: warning: unused parameter 'ctx' [-Wunused-parameter]
 13461 | void llama_split_layers_weighted(struct llama_context * ctx, float device_weights[], size_t num_weights) {
       |                                                         ^
llama.cpp:13461:68: warning: unused parameter 'device_weights' [-Wunused-parameter]
 13461 | void llama_split_layers_weighted(struct llama_context * ctx, float device_weights[], size_t num_weights) {
       |                                                                    ^
llama.cpp:13461:93: warning: unused parameter 'num_weights' [-Wunused-parameter]
 13461 | void llama_split_layers_weighted(struct llama_context * ctx, float device_weights[], size_t num_weights) {
       |                                                                                             ^
llama.cpp:14475:5: warning: no previous prototype for function 'llama_process_mpi_transaction' [-Wmissing-prototypes]
 14475 | int llama_process_mpi_transaction(
       |     ^
llama.cpp:14475:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
 14475 | int llama_process_mpi_transaction(
       | ^
       | static 
llama.cpp:14512:13: warning: 'break' will never be executed [-Wunreachable-code-break]
 14512 |             break;
       |             ^~~~~
llama.cpp:14487:13: warning: 'break' will never be executed [-Wunreachable-code-break]
 14487 |             break;
       |             ^~~~~
llama.cpp:14522:13: warning: unused variable 'count' [-Wunused-variable]
 14522 |     int32_t count;
       |             ^~~~~
llama.cpp:14517:5: warning: no previous prototype for function 'llama_process_mpi_worker' [-Wmissing-prototypes]
 14517 | int llama_process_mpi_worker(
       |     ^
llama.cpp:14517:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
 14517 | int llama_process_mpi_worker(
       | ^
       | static 
llama.cpp:14532:13: warning: 'break' will never be executed [-Wunreachable-code-break]
 14532 |             break;
       |             ^~~~~
llama.cpp:14537:13: warning: 'break' will never be executed [-Wunreachable-code-break]
 14537 |             break;
       |             ^~~~~
llama.cpp:14550:13: warning: 'break' will never be executed [-Wunreachable-code-break]
 14550 |             break;
       |             ^~~~~
2 warnings generated.
13 warnings generated.
```

AutonomicPerfectionist commented:
Interesting... At the moment the wrapper caches are the only parts using maps, but those caches are not actually being used at the moment. I'll add the include in my next push if I end up needing the caches, otherwise I'll remove them and therefore the dependency on map

LeaveNhA commented:
@AutonomicPerfectionist, may I ask, is it to be expected to get crash with MPS backend with ngl parameter setted with > 1?

AutonomicPerfectionist commented:
Yeah, that's pretty much expected, I haven't tested with anything but CPU as the wrapped backend. So metal, CUDA, Vulkan, SYCL, I expect to all crash horrifically. The plan is to fix that so you can run with any backend. If you can, could you put all the details of what you tried here so I can fix it? I only have access to NVIDIA and AMD GPUs so if you have a Mac with Metal it would be a big help to provide any error messages. Also try to build with debug mode on, and in the case of a SEGFAULT you can either use a debugger to investigate further or decompile with `objdump` and search for the program offsets that trigger the SEGFAULT.

LeaveNhA commented:
Yes, of course. I will inform you when I get a chance to test it. Probably, today afternoon.

AutonomicPerfectionist committed:
Fix non-CPU backend wrapping
